Unity-Technologies/ScriptableRenderPipeline
# Deleted SG Asset Dialog

Implementation of this old card "_Do Something (Anything) when the Shader Graph file is deleted while the window into the graph is still open_":
https://favro.com/organization/c564ede4ed3337f7b17986b6/d06b8de3e40da6d340b5b063?card=Uni-49784

![DeletionSaveAsMenu](https://user-images.githubusercontent.com/3393041/70601826-d1b4d900-1ba7-11ea-8299-a5e3cbf92dcc.gif)

---
**General Notes**
- Stops users from entering in a still open Shader Graph that is in a buggy soft lock mode; they either ought to close the Shader Graph or last chance to Save As.
- Works with all graphs, allowing users a last chance Save As if they regret the deletion of the Shader Graph
- This popup does not happen until the window has been selected.
- Note that "Save As" is the secondary, not the primary, option. Please feel free to give feedback on that. Although that's abnormal as far as saving stuff goes, I did that because I figured that primary should be the default use case, after all the user will have deleted the file on purpose 96% of the time.

---
**Notes to QA**
- ^ stuff above
- Should work with various window arrangements
- If you have the bug where Save As opens in the wrong project, that should be fixed in another PR of mine. Here is a combined branch to test that: sg/combined/deletion-dialog-menu-navigation
- More ways to delete stuff you care to test? Simple script that deletes Assets? That would be more testing unity's asset deletion detection system, but might be good nonetheless if you care to do that.
- V See manually tested.

---
**Notes to Docs:**
Since this has a new Dialog popup, is the wording okay?:
_Note:_ this has been changed since PR has been first created.

"_____ Graph Asset Missing!"
"_Path-to-file-&-file-name_ has been deleted or moved outside of Unity."
_space_
"What do you wish to do?"
1) "Close Window"
2) "Save As"
<img width="429" alt="Screen Shot 2019-12-20 at 5 40 42 PM" src="https://user-images.githubusercontent.com/3393041/71300205-f51e1780-234f-11ea-999a-9a92d59bd611.png">

---
**Testing:**

Manually Tested:
- Delete -> Don't Save (Close), Delete -> Save As, Delete -> Save As the same file, many times and ways
- Deleting inside of Unity and outside of Unity via File Explorer
- Shader Graphs and Sub Graphs
- Various window / tab arrangements. The popup (should) only popup once the tab or Shader Graph window has been selected, as opposed to immediately when the file has been deleted.

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline/tree/sg%252Fdeletion-dialog/.yamato%252Fupm-ci-packages.yml%2523all_package_ci_fast-2020.1/1539611/job


https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline/tree/sg%252Fdeletion-dialog/.yamato%252Fupm-ci-shadergraph.yml%2523All_ShaderGraph_trunk
(Master compare: https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline/tree/master/.yamato%252Fupm-ci-packages.yml%2523all_package_ci/1009336/job)
